### ISSUE TITLE
Problems About Memory Monitoring

### ISSUE THREAD

**Issue by MrLi2018** [javadelight/delight-nashorn-sandbox#125]
In the following case, an infinite loop is performed, but I do not know why the memory usage increases sharply.

**Comment by MrLi2018** [comment]
![9EC7DE15-69BE-4D03-8FF5-BD73FFC41F6C](https://user-images.githubusercontent.com/39287657/156880902-66952fc6-eb56-4055-93a1-b8b72ceb3903.png)

**Comment by MrLi2018** [comment]
![B2C04E58-09CE-4DAB-914D-4C4E08CEDF57](https://user-images.githubusercontent.com/39287657/156880932-74ce3824-78a9-4bc6-9489-22f664cac966.png)

**Comment by mxro** [comment]
Thansk for raising this issue!

Interesting. I would assume the sandbox will add the interrupt function and then call that for every loop run.

```
var \__it = Java.type('delight.nashornsandbox.internal.InterruptTest');var \__if=function(){\__it.test();};
```

You can double check what JS is generated by setting the following property `log4j.logger.delight.nashornsandbox.NashornSandbox=DEBUG`.

I don't think so that this should lead to a memory leak. Maybe it is Nashorn internals that allocate more objects?

**Comment by MrLi2018** [comment]
Thank you for your answer, I'm not saying it caused a memory leak, but I'm asking why the memory increases has anything to do with this injection method, which seems to be very simple.

**Comment by MrLi2018** [comment]
Because now I find that some js code that shouldn't have gone out of memory is now out, but I'm not sure how I should analyze it.

**Comment by mxro** [comment]
Yes I don't see anything in the code here that could cause a memory leak. I would assume it is some allocation that Nashorn may make upon every method invocation.

But if you see anything in the code generated by the sandbox that looks like it could cause a memory leak, please let me know and happy to patch it then!

**Comment by MrLi2018** [comment]
hreadMonitor.registerThreadToMonitor(Thread.currentThread()) 
  synchronized(this.monitor) {
                if (this.threadToMonitor == null) {
                    this.monitor.wait((this.maxCPUTime + 500L) / 1000000L);
                }

                if (this.threadToMonitor == null) {
                    this.timedOutWaitingForThreadToMonitor = true;
                    throw new IllegalStateException("Executor thread not set after " + this.maxCPUTime / 1000000L + " ms");
                }
            }
Hello, I'd like to ask if this.threadToMonitor may be empty if Thread.currentThread is in the interrupt state.

**Comment by MrLi2018** [comment]
Hello mxro, I find a strange phenomenon. My total heap memory is only 800 MB, but the sandbox calculates 3488 MB. I don't know why the error is so large.

![81ACD530-3DFD-4694-BA93-07EF16260B54](https://user-images.githubusercontent.com/39287657/160235656-3f140dda-ba95-4e69-a912-ec71f2c8b539.png)

![DE3583CF-9517-4EE4-A944-6544C87947CC](https://user-images.githubusercontent.com/39287657/160235721-c9202585-77f9-4530-a1c0-b5d20bb71090.png)

![3D926927-49F2-4907-A264-D8EE6C4D91DD](https://user-images.githubusercontent.com/39287657/160235739-63ac759d-5ed3-488a-85f6-b92e53946972.png)
@mxro

**Comment by mxro** [comment]
Could this be related to the stage memory https://github.com/javadelight/delight-nashorn-sandbox/blob/73b8a92a87c2e69c54e24a48bd1ca420d976ef76/src/main/java/delight/nashornsandbox/internal/ThreadMonitor.java#L65 

Otherwise the Sandbox doesn't do much else than just read the memory through the bean: https://github.com/javadelight/delight-nashorn-sandbox/blob/73b8a92a87c2e69c54e24a48bd1ca420d976ef76/src/main/java/delight/nashornsandbox/internal/ThreadMonitor.java#L205 - assuming that's not the most accurate to begin with?

**Comment by MrLi2018** [comment]
It can't be related to stage memory. I used version 0.2.0.

**Comment by MrLi2018** [comment]
After my thread runs for a long time, the memory value suddenly increases at some point. I suspect that the memory calculation thread becomes inaccurate after running for a long time.

**Comment by MrLi2018** [comment]
[delight-nashorn-sandbox/src/main/java/delight/nashornsandbox/internal/ThreadMonitor.java](https://github.com/javadelight/delight-nashorn-sandbox/blob/73b8a92a87c2e69c54e24a48bd1ca420d976ef76/src/main/java/delight/nashornsandbox/internal/ThreadMonitor.java#L65)

Line 65 in [73b8a92](https://github.com/javadelight/delight-nashorn-sandbox/commit/73b8a92a87c2e69c54e24a48bd1ca420d976ef76)

 		// if maxMemory is larger than 100M, split heap memory count into 4 stage, and offset will be 2


This problem also occurs when threads whose memory is less than 100 MB run for a long time. You are advised to modify this problem.

**Comment by MrLi2018** [comment]
https://stackoverflow.com/questions/31338441/why-is-getthreadallocatedbytes-not-consistent

**Comment by mxro** [comment]
So what would be the easiest solution?

**Comment by MrLi2018** [comment]
I'm not sure at present. I just set the memory limit to 100 MB and used the segment calculation, but I found that the error still exists just to avoid it.
